Don't reserve vector capacity <= 0 in ISOTrackEncryptionBox::parsePayload

chirags27 opened the pull request:
#### 211a030a510d221c9286ee7a4166dddf79e352e0
<pre>
Don&apos;t reserve vector capacity &lt;= 0 in ISOTrackEncryptionBox::parsePayload
<a href="https://bugs.webkit.org/show_bug.cgi?id=258156">https://bugs.webkit.org/show_bug.cgi?id=258156</a>
rdar://110421682

Reviewed by Andy Estes.

This change fixes an issue where we read something which is &lt; 0 in an
int8_t and then pass it to a size_t, which results in a crash because we
attempt to allocate a huge buffer to due sign conversion.

* Source/WebCore/platform/graphics/iso/ISOTrackEncryptionBox.cpp:
(WebCore::ISOTrackEncryptionBox::parsePayload):

Canonical link: <a href="https://commits.webkit.org/265227@main">https://commits.webkit.org/265227@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/eedb13b1c1ab8823b4be2efcfd8c58aad0cbff08

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10270 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10507 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10772 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11916 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/9886 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10282 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12502 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10459 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12834 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10428 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11159 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8651 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12302 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8466 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9297 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16581 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9565 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9447 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/12698 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9904 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8035 "3 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9063 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2471 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13313 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9742 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/platform/graphics/iso/ISOTrackEncryptionBox.cpp:
```diff
@@ -87,7 +87,7 @@ bool ISOTrackEncryptionBox::parsePayload(DataView& view, unsigned& offset)
     if (m_defaultIsProtected == 1 && !m_defaultPerSampleIVSize) {
         int8_t defaultConstantIVSize = 0;
-        if (!checkedRead<int8_t>(defaultConstantIVSize, view, offset, BigEndian))
+        if (!checkedRead<int8_t>(defaultConstantIVSize, view, offset, BigEndian) || defaultConstantIVSize <= 0)
```
Seems like this unnecessarily changes behavior when `defaultConstantIVSize` is 0. Should we do this instead?
```suggestion
        if (!checkedRead<int8_t>(defaultConstantIVSize, view, offset, BigEndian) || defaultConstantIVSize < 0)
```

chirags27 committed:
Don't reserve vector capacity <= 0 in ISOTrackEncryptionBox::parsePayload
https://bugs.webkit.org/show_bug.cgi?id=258156
rdar://110421682

Reviewed by Andy Estes.

This change fixes an issue where we read something which is < 0 in an
int8_t and then pass it to a size_t, which results in a crash because we
attempt to allocate a huge buffer to due sign conversion.

* Source/WebCore/platform/graphics/iso/ISOTrackEncryptionBox.cpp:
(WebCore::ISOTrackEncryptionBox::parsePayload):

Canonical link: https://commits.webkit.org/265227@main
